contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,209 @@
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+#[derive(Debug)]
+pub enum DatabaseConfigError<'a> {
+    MissingConfig,
+    MissingTable(&'a str),
+    MissingConnectionString(&'a str),
+    InvalidPoolSize(i64),
+}
+
+impl<'a> DatabaseConfigError<'a> {
+    pub fn pretty_print(&self) {
+        error!("{}", self);
+    }
+}
+
+impl<'a> Display for DatabaseConfigError<'a> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            DatabaseConfigError::MissingConfig => {
+                write!(f, "Could not find the `databases` key in the configuration")
+            },
+            DatabaseConfigError::MissingTable(table_name) => {
+                write!(f, "Could not find the configuration named `{}`", table_name)
+            },
+            DatabaseConfigError::MissingConnectionString(conn_string) => {
+                write!(f, "The specified connection string, `{}`, is invalid", conn_string)
+            },
+            DatabaseConfigError::InvalidPoolSize(pool_size) => {
+                write!(f, "The specified pool size, `{}`, is invalid", pool_size)
+            },
+        }
+    }
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(
+    name: &'a str,
+    from: &'a config::Config
+) -> Result<DatabaseConfig<'a>, DatabaseConfigError<'a>> {
+    // Find the first `databases` config that's a table with a key of 'name'
+    // equal to `name`.
+    let connection_config = from.get_slice("databases")
+        .map_err(|_| DatabaseConfigError::MissingConfig)?
+        .iter()
+        .filter_map(|config| config.as_table())
+        .filter(|table| table.get("name").map(|n| n.as_str() == Some(name)).unwrap_or(false))
+        .next()
+        .ok_or(DatabaseConfigError::MissingTable(name))?;
+
+    let url = connection_config.get("url")
+        .and_then(Value::as_str)
+        .ok_or(DatabaseConfigError::MissingConnectionString("Could not fetch a valid connection string for the database configuration"))?;
+
+    let pool_size = connection_config.get("pool_size")
+        .and_then(Value::as_integer)
+        .unwrap_or(from.workers as i64);
+
+    if pool_size < 1 {
+        return Err(DatabaseConfigError::InvalidPoolSize(pool_size));
+    }
+
+    Ok(DatabaseConfig { url, pool_size, extras: connection_config })
+}
+
+/// Returns an iterator containing the names of all the possible database connections.
+/// The primary use within `rocket_contrib` for this method is in relation to calling out
+/// valid database connections that are specified in the `Rocket.toml` file.
+pub fn available_database_configs<'a>(config: &'a config::Config) -> Result<impl Iterator<Item=&'a str>, DatabaseConfigError<'a>> {
```

Comment:
Truth be told, this method is never used by us and is only supplied to the user for their purposes. We can axe it or keep it. The requirement of having a `config::Config` sent in as an argument makes it difficult to use this for pretty printed error messages.

Reply:
If we don't use it, scrap it.